cheri-riscv: csetequalexact

nwf opened the pull request:
Is there a better way than trying to enumerate all the fields of the decoded structure like this?

nwf committed:
cheri-riscv: csetequalexact

arichardson reviewed (changes requested):
Can we use this for cexeq? I think the new code is probably better than the existing MIPS version.

nwf committed:
Refactor cap_exactly_equal out to cheri-compressed-cap

Wrap in cheri-common and use in both MIPS and RISC-V helpers

nwf commented:
Like that?

arichardson commented on target/mips/op_helper_cheri.c:
```diff
@@ -632,37 +630,18 @@ target_ulong CHERI_HELPER_IMPL(cleu(CPUArchState *env, uint32_t cb, uint32_t ct)
     const cap_register_t *cbp = get_readonly_capreg(env, cb);
     const cap_register_t *ctp = get_readonly_capreg(env, ct);
-    return (target_ulong) cap_exactly_equal(cbp, ctp) ? false : true;
+    return !!cap_exactly_equal(cbp, ctp) ? false : true;
```
```suggestion
    return cap_exactly_equal(cbp, ctp) ? false : true;
```

nwf replied:
Whoop, yep!  Over-hasty copy-paste.

arichardson commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -880,6 +880,15 @@ target_ulong CHERI_HELPER_IMPL(ctestsubset(CPUArchState *env, uint32_t cb,
+target_ulong CHERI_HELPER_IMPL(csetequalexact(CPUArchState *env, uint32_t cb,
+                                              uint32_t ct))
+{
+    const cap_register_t *cbp = get_capreg_0_is_ddc(env, cb);
```
I am unsure what is more useful: comparing to CNULL or to DDC. Considering we don't use 0 == ddc very much for RISC-V I'd be tempted to use CNULL (same for the sail PR). This would also match MIPS.

But I don't really mind either way. @jrtc27 ?

jrtc27 replied:
Using DDC for 0 has always been weird, and IMO should be reserved for cases where CNULL is meaningless but DDC provides useful behaviour (ie CToPtr, CFromPtr and CBuildCap), or where CNULL is unlikely to be very useful (ie CTestSubset). Both could conceivably be useful here, but CNULL is less surprising and since it's a meaningful operation to perform I think we should use CNULL.

nwf replied:
<s>I just stole this from PeterRugg's implementation in Flute.</s> Apparently not; sorry for the mis-attribution!

It does have the sort of "cute" result that

  * `CSetEqualExact rd, 0, 0` compares DDC and NULL
  * `CSetEqualExact rd, cN, 0` compares cN against NULL
  * `CSetEqualExact rd, 0, cN` compares cN against DDC

The remaining diagonal entries are fairly useless, tho', so maybe the "0, 0" one should be too.  (And the redundancy in the encodings also makes me unhappy, but I guess "half of the capreg*capreg space" isn't very many actual opcodes.)

jrtc27 commented on target/mips/op_helper_cheri.c:
```diff
@@ -632,37 +630,18 @@ target_ulong CHERI_HELPER_IMPL(cleu(CPUArchState *env, uint32_t cb, uint32_t ct)
     const cap_register_t *cbp = get_readonly_capreg(env, cb);
     const cap_register_t *ctp = get_readonly_capreg(env, ct);
-    return (target_ulong)cap_exactly_equal(cbp, ctp);
+    return !!cap_exactly_equal(cbp, ctp);
```
Um, why? `cap_exactly_equal` already returns `bool`.

nwf replied:
Habit for coercing things into {0, 1}.

jrtc27 commented on target/mips/op_helper_cheri.c:
```diff
@@ -632,37 +630,18 @@ target_ulong CHERI_HELPER_IMPL(cleu(CPUArchState *env, uint32_t cb, uint32_t ct)
     const cap_register_t *cbp = get_readonly_capreg(env, cb);
     const cap_register_t *ctp = get_readonly_capreg(env, ct);
-    return (target_ulong) cap_exactly_equal(cbp, ctp) ? false : true;
+    return cap_exactly_equal(cbp, ctp) ? false : true;
```
... or just `!`.

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -107,6 +107,19 @@ static inline int64_t cap_get_otype(const cap_register_t* c) {
 
+static inline bool cap_exactly_equal(const cap_register_t *cbp, const cap_register_t *ctp)
+{
+    return !!(cbp->cr_tag           == ctp->cr_tag           &&
```
Why?! Everything in here is a bool already.

nwf replied:
Copy-paste.

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -107,6 +107,19 @@ static inline int64_t cap_get_otype(const cap_register_t* c) {
+static inline bool cap_exactly_equal(const cap_register_t *cbp, const cap_register_t *ctp)
+{
+    return !!(cbp->cr_tag           == ctp->cr_tag           &&
+              cap_get_cursor  (cbp) == cap_get_cursor  (ctp) &&
```
I'm not a fan of this formatting

nwf replied:
What would you prefer as an alternative?

jrtc27 commented on disas/riscv.c:
```diff
@@ -1262,6 +1263,7 @@ const rv_opcode_data opcode_data[] = {
     [rv_op_ccopytype] = { "ccopytype", rv_codec_r, rv_fmt_cd_cs1_cs2, NULL, 0, 0, 0 },
     [rv_op_ccseal] = { "ccseal", rv_codec_r, rv_fmt_cd_cs1_cs2, NULL, 0, 0, 0 },
     [rv_op_ctestsubset] = { "ctestsubset", rv_codec_r, rv_fmt_rd_cs1_cs2, NULL, 0, 0, 0 },
+    [rv_op_csetequalexact] = { "csetequalexact", rv_codec_r, rv_fmt_rd_cs1_cs2, NULL, 0, 0, 0 },
```
`cseqx` everywhere please to match LLVM and `crrl`/`cram`

arichardson commented on target/cheri-common/cheri_utils.h:
```diff
@@ -107,6 +107,19 @@ static inline int64_t cap_get_otype(const cap_register_t* c) {
+static inline bool cap_exactly_equal(const cap_register_t *cbp, const cap_register_t *ctp)
+{
+    return cbp->cr_tag           == ctp->cr_tag           &&
+           cap_get_cursor  (cbp) == cap_get_cursor  (ctp) &&
+           cap_get_base    (cbp) == cap_get_base    (ctp) &&
+           cap_get_length65(cbp) == cap_get_length65(ctp) &&
+           cap_get_otype   (cbp) == cap_get_otype   (ctp) &&
```
```suggestion
           cap_get_cursor(cbp)   == cap_get_cursor(ctp)   &&
           cap_get_base(cbp)     == cap_get_base(ctp)     &&
           cap_get_length65(cbp) == cap_get_length65(ctp) &&
           cap_get_otype(cbp)    == cap_get_otype(ctp)    &&
```

Personally I would't bother aligning the equals, but I don't mind either way. I'd prefer not to have spaces between function name and parens though.

jrtc27 commented:
Technically this isn't quite right as it relies on both operands being in canonical form, which _should_ be architecturally enforced for tagged quantities but for untagged quantities you can feed in whatever junk you like. Architecturally it needs to be an exact equality check on the CLEN+1-bit value, so the easiest thing might just be to encode the capabilities and compare that? But looking at the exponent is probably enough to fix this? Though we've managed to get by with a buggy implementation on MIPS here (didn't even look at cr_uperms, let alone the reserved fields and what exponent was chosen)...

nwf commented:
Oof.  Rudely, `cheri-compressed-cap` exports `operator==` on its capability types... only for C++ consumers.  I will push that out to C and then use that instead of recreating it in qemu.

In practice the revoker only exercises "at least one operand tagged" paths of this instruction, so it's not very surprising that there's bugs on the "neither tagged" paths.  The missing `cr_uperms` is a little more worrying, but again, the only user was the revoker and it will never restore cleared `cr_uperms` bits (and is insensitive to them when testing).

nwf commented:
If this looks reasonable, we should `subrepo push cheri-compressed-cap`, I guess?

jrtc27 commented on target/cheri-common/cheri_utils.h:
```diff
@@ -107,6 +107,18 @@ static inline int64_t cap_get_otype(const cap_register_t* c) {
     return result < CAP_LAST_NONRESERVED_OTYPE ? result : result - CAP_MAX_REPRESENTABLE_OTYPE - 1;
 }
 
+#ifdef CHERI_128
+static inline bool cap_exactly_equal(const cap_register_t *cbp, const cap_register_t *ctp)
+{
+    return cc128_exactly_equal(cbp, ctp);
+}
+#else
+static inline bool cap_exactly_equal(const cap_register_t *cbp, const cap_register_t *ctp)
+{
+    return cc256_exactly_equal(cbp, ctp);
+}
+#endif
+
```
```suggestion
static inline bool cap_exactly_equal(const cap_register_t *cbp, const cap_register_t *ctp)
{
#ifdef CHERI_128
    return cc128_exactly_equal(cbp, ctp);
#else
    return cc256_exactly_equal(cbp, ctp);
#endif
}

```

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -880,6 +880,15 @@ target_ulong CHERI_HELPER_IMPL(ctestsubset(CPUArchState *env, uint32_t cb,
     return (target_ulong)is_subset;
 }
 
+target_ulong CHERI_HELPER_IMPL(cseqx(CPUArchState *env, uint32_t cb,
+                                     uint32_t ct))
+{
+    const cap_register_t *cbp = get_readonly_capreg(env, cb);
+    const cap_register_t *ctp = get_readonly_capreg(env, ct);
+
+    return cap_exactly_equal(cbp, ctp);
+}
```
Hm. I don't know if this should live here or in target/riscv, given target/mips has its own cexeq instead...

arichardson replied:
Since the actual comparison has been factored out, I'd move this to the RISCV helpers.

nwf replied:
I'm tempted to make `target/mips` translate `cexeq` to `cseqx`, since it's the same code?

arichardson replied:
Yes that would probably be even better. I guess we could de-Riscv-ify the name of the helper though. I prefer readable names to weird abbreviations.

arichardson replied:
Ah but that would make the translate code awkward. Let's just keep the RISCV name then

arichardson commented on target/cheri-common/cheri-compressed-cap/cheri_compressed_cap_common.h:
```diff
@@ -102,16 +105,26 @@ struct _cc_N(cap) {
-               cr_perms == other.cr_perms && cr_uperms == other.cr_uperms && cr_otype == other.cr_otype &&
-               cr_ebt == other.cr_ebt && cr_flags == other.cr_flags && cr_reserved == other.cr_reserved &&
-               cr_tag == other.cr_tag;
+      return _cc_N(exactly_equal)(this, other);
```
I'd move the implementation out of line to avoid the forward declaration. I think there's another ifdef c++ further down where this could go. But I don't really mind this is fine too.

arichardson commented on target/cheri-common/cheri-compressed-cap/cheri_compressed_cap_common.h:
```diff
@@ -102,16 +105,26 @@ struct _cc_N(cap) {
+         a->cr_otype == b->cr_otype &&
+         a->cr_ebt == b->cr_ebt &&
+         a->cr_flags == b->cr_flags &&
+         a->cr_reserved == b->cr_reserved;
```
We should check that there isn't any padding so the compiler can simplify these comparisons. But since it's not in the critical path this really doesn't matter.

arichardson commented on target/cheri-common/cheri_utils.h:
```diff
@@ -107,6 +107,15 @@ static inline int64_t cap_get_otype(const cap_register_t* c) {
+#ifdef CHERI_128
+    return cc128_exactly_equal(cbp, ctp);
+#else
+    return cc256_exactly_equal(cbp, ctp);
```
Seems fine to me but we could probably also drop the 256 case (and add a 64 case later)

jrtc27 replied:
Eh, easy enough to keep.

nwf committed:
CHERI MIPS: remove cexeq helper in favor of cseqx instead

While "cseqx" may not be the most ideal name for op_helper_cheri_common.c, it
would be awkward to make the RISC-V translation logic use a different name.

nwf commented:
Alarmingly, this is breaking some `cbuildcap` test programs because the `cr_ebt` fields differ across round-tripping `ccleartag; cbuildcap`.

jrtc27 commented:
> Alarmingly, this is breaking some `cbuildcap` test programs because the `cr_ebt` fields differ across round-tripping `ccleartag; cbuildcap`.

Congratulations you are now the proud owner of a `cbuildcap` bug :)

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -450,6 +450,7 @@ void CHERI_HELPER_IMPL(cbuildcap(CPUArchState *env, uint32_t cd, uint32_t cb,
         result.cr_perms = ctp->cr_perms;
         result.cr_uperms = ctp->cr_uperms;
         result._cr_cursor = ctp->_cr_cursor;
+        result.cr_ebt = ctp->cr_ebt;
```
Missing flags and reserved still. But at this point why are we even using the authorising capability? Why not not just `cap_register_t result = *ctp; if (!cap_is_sealed_entry(ctp)) result.cr_otype = CAP_OTYPE_UNSEALED;` (though that's wrong on RISC-V, we don't want to unseal arbitrary sealed things, it only permits sentries and unsealed caps, always preserving the otype). There should be literally nothing inherited from the authorising capability other than architectural provenance.

nwf replied:
Glancing at the riscv sail, I don't see any conditioning on the `otype` beyond "is a sentry" (and if so, return a sentry, otherwise, return unsealed) (though it's missing the check for permit_execute when it does decide to return a sentry...).

jrtc27 replied:
Sail is wrong; it has `  } else if cb_val.sealed then {  handle_cheri_reg_exception(CapEx_SealViolation, cb);` further up, and is also rederiving. Flute has the "right" implementation: https://github.com/CTSRD-CHERI/Flute/blob/2a8d8708c77e477e9be5d4938b4d825521de369f/src_Core/CPU/EX_ALU_functions.bsv#L1713-L1736

jrtc27 replied:
Although I have now spotted a bug in the cap idx Flute reports on exception for register 0... :)

nwf replied:
I believe both of the following statements, but both tentatively: `cb_val` is the authority, and it's correct to reject sealed authorities in `cbuildcap`?

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -439,21 +439,23 @@ void CHERI_HELPER_IMPL(cbuildcap(CPUArchState *env, uint32_t cd, uint32_t cb,
+        cap_register_t result = *ctp;
+
+        /*
+	 * cbuildcap is allowed to seal at any ambiently-available otype,
+	 * subject to their construction conditions.
```
Whitespace. Also please separate out this diff, it's not relevant to csetequalexact other than that you found a bug with cbuildcap when testing it.

jrtc27 commented on target/mips/op_helper_cheri.c:
```diff
@@ -632,37 +630,11 @@ target_ulong CHERI_HELPER_IMPL(cleu(CPUArchState *env, uint32_t cb, uint32_t ct)
-    return (target_ulong)cap_exactly_equal(cbp, ctp);
-}
-
 target_ulong CHERI_HELPER_IMPL(cnexeq(CPUArchState *env, uint32_t cb, uint32_t ct))
```
I wonder if it'd be better to delete the negated helper, instead doing the negation in TCG.

arichardson replied:
I guess could theoretically might be marginally slower since tcg does slightly more work.
However we add a single `tcg_gen_xori_i64(t0, t0, 1)` line and delete a 6 line helper+declaration, so sounds good to me.

nwf committed:
MIPS: replace cnexeq helper with in-TCG negation of cseqx

arichardson commented on target/cheri-common/cheri-compressed-cap/cheri_compressed_cap_common.h:
```diff
@@ -101,17 +101,24 @@ struct _cc_N(cap) {
     inline uint32_t permissions() const { return cr_perms; }
     inline uint32_t type() const { return cr_otype; }
     inline bool is_sealed() const { return cr_otype != _CC_N(OTYPE_UNSEALED); }
-    inline bool operator==(const _cc_N(cap) & other) const {
```
The declaration is still needed.

arichardson reviewed (approved):
LGTM other than the one comment.

arichardson commented:
We should probably merge the buildcap change first to keep Jenkins happy.

brooksdavis commented:
It would be good to get this merged so Galois can build a new qemu for the FETT docker images.

jrtc27 commented:
I don’t think there are any outstanding blockers for this; @arichardson?
